src/main/java/com/epam/spring/library/service/impl/BookServiceImpl.java

```diff
@@ -0,0 +1,112 @@
+
+    @Override
+    public BookDTO getBook(String isbn) {
+        return mapper.toDTO(repository.getBook(isbn));
+    }
+
+    @Override
+    public List<BookDTO> getAllBooks() {
+        return mapper.toDTO(repository.getAllBooks());
+    }
+
+    @Override
+    public BookDTO createBook(BookDTO bookDTO) {
+        Book newBook = mapper.toBook(bookDTO);
+        return mapper.toDTO(repository.createBook(newBook));
+    }
+
+    @Override
+    public BookDTO updateBook(String isbn, BookDTO bookDTO) {
+        Book toUpdate = repository.getBook(isbn);
+        mapper.updateBook(bookDTO, toUpdate);
+        return mapper.toDTO(repository.updateBook(isbn, toUpdate));
+    }
+
+    @Override
+    public void deleteBook(String isbn) {
+        repository.deleteBook(isbn);
+    }
+
+    @Override
+    public Set<AuthorDTO> getBookAuthors(String isbn) {
+        Book book = repository.getBook(isbn);
+        return authorMapper.toDTO(book.getAuthors());
+    }
+
+    @Override
+    public Set<AuthorDTO> setAuthorsOfBook(String isbn,
+                                           List<AuthorDTO> authorDTOs) {
+        Book book = repository.getBook(isbn);
+        List<Author> authors = getNewAuthorList(authorDTOs);
+        if (!book.getAuthors().isEmpty()) {
+            throw new RuntimeException(
+                    "author list of book isbn#" + isbn + " is not " + "empty");
+        }
+        book.getAuthors().addAll(authors);
+        return authorMapper.toDTO(book.getAuthors());
+    }
+
+    @Override
+    public Set<AuthorDTO> updateAuthorsOfBook(String isbn,
+                                              List<AuthorDTO> authorDTOs) {
+        Book book = repository.getBook(isbn);
+        List<Author> newAuthorList = getNewAuthorList(authorDTOs);
+        book.getAuthors().clear();
+        book.getAuthors().addAll(newAuthorList);
+        return authorMapper.toDTO(book.getAuthors());
+    }
+
+    private List<Author> getNewAuthorList(List<AuthorDTO> authorDTOs) {
+        return authorDTOs.stream()
+                         .map(AuthorDTO::getDefaultName)
+                         .map(authorRepository::getAuthor)
+                         .collect(Collectors.toList());
+    }
+
+    @Override
+    public void clearAuthorListOfBook(String isbn) {
+        Book book = repository.getBook(isbn);
+        book.getAuthors().clear();
+    }
+
+    @Override
+    public void deleteAuthorOfBook(String isbn, String authorName) {
+        Book book = repository.getBook(isbn);
+        Author author = authorRepository.getAuthor(authorName);
+        book.getAuthors().remove(author);
+    }
+
+    @Override
+    public Set<AuthorDTO> addAuthorsToBook(String isbn,
+                                           List<AuthorDTO> authorDTOs) {
+        Book book = repository.getBook(isbn);
+        List<Author> authors = getNewAuthorList(authorDTOs);
+        book.getAuthors().addAll(authors);
+        return authorMapper.toDTO(book.getAuthors());
```

Comment:
get rid of local variables

Reply:
I agree with the "authors" variable redundancy, but I have a concern about the "book" variable. If I get rid of it, I will make 2 calls to my database and, no matter is it ListImpl or real DB, I consider them as "heavy" calls. Shouldn't I keep it of such consideration?

Reply:
Looking at the code I assume (with high degree of probability) that you are going to use Spring Data JPA. If so - lets keep it as is and examine how the first level cache (which is Persistence Context) works and if it prevents application from unnecessary calls to the database